File: samples/cxx_samples/translation_samples.cpp
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

#include "stdafx.h"
#include <string>
#include <vector>

#include <speechapi_cxx.h>


using namespace std;
using namespace Microsoft::CognitiveServices::Speech::Recognition;
using namespace Microsoft::CognitiveServices::Speech::Recognition::Translation;


// Translation with microphone input.
// <TranslationWithMicrophone>
void TranslationWithMicrophone()
{
    // Gets recognizer factory.
    auto factory = RecognizerFactory::GetDefault();

    // Sets subscription key
    // Replaces with your own subscription key.
    factory->SetSubscriptionKey(L"YourSubscriptionKey");

    // Sets source and target languages
    auto fromLanguage = L"en-US";
    vector<wstring> toLanguages { L"de-DE", L"fr-FR" };

    // Creates a translation recognizer using microphone as audio input.
    auto recognizer = factory->CreateTranslationRecognizer(fromLanguage, toLanguages);
    wcout << L"Say something...\n";

    // Starts translation. It returns when the first utterance has been recognized.
    auto result = recognizer->RecognizeAsync().get();

    // Checks result.
    if (result->Reason != Reason::Recognized)
    {
        wcout << L"There was an error in speech recognition, reason " << int(result->Reason) << L"-" << result->Text << '\n';
    }
    else if (result->TranslationStatus != TranslationStatus::Success)
    {
        wcout << L"There was an error in translation, status: " << int(result->TranslationStatus) << '\n';
    }
    else
    {
        wcout << "We recognized in " << fromLanguage << ":" << result->Text << '\n';
        for (const auto& it : result->Translations)
        {
            wcout << L"    Translated into " << it.first.c_str() << ":" << it.second.c_str();
        }
    }
}
// </TranslationWithMicrophone>

// Translation with file input.
// <TranslationWithFile>
void TranslationWithFile()
{
    // Gets recognizer factory.
    auto factory = RecognizerFactory::GetDefault();

    // Sets subscription key
    // Replaces with your own subscription key.
    factory->SetSubscriptionKey(L"YourSubscriptionKey");

    // Sets source and target languages
    auto fromLanguage = L"en-US";
    vector<wstring> toLanguages{ L"de-DE", L"fr-FR" };

    // Creates a translation recognizer using file as audio input.
    // Replaces with your own audio file name.
    auto recognizer = factory->CreateTranslationRecognizerWithFileInput(L"YourAudioFileName", fromLanguage, toLanguages);

    // Starts translation. It returns when the first utterance has been recognized.
    auto result = recognizer->RecognizeAsync().get();

    // Checks result.
    if (result->Reason != Reason::Recognized)
    {
        wcout << L"There was an error in speech recognition, reason " << int(result->Reason) << L"-" << result->Text << '\n';
    }
    else if (result->TranslationStatus != TranslationStatus::Success)
    {
        wcout << L"There was an error in translation, status: " << int(result->TranslationStatus) << '\n';
    }
    else
    {
        wcout << "We recognized in " << fromLanguage << ":" << result->Text << '\n';
        for (const auto& it : result->Translations)
        {
            wcout << L"    Translated into " << it.first.c_str() << ":" << it.second.c_str();
        }
    }
}
// </TranslationWithFile>

// Defines event handlers for different events.
// <TranslationContinuousRecognitionUsingEvents>
static void OnPartialResult(const TranslationTextResultEventArgs& e)
{
    wcout << L"IntermediateResult: Recognized text:" << e.Result.Text << '\n';
    for (const auto& it : e.Result.Translations)
    {
        wcout << L"    Translated into " << it.first.c_str() << ":" << it.second.c_str();
    }
}

static void OnFinalResult(const TranslationTextResultEventArgs& e)
{
    wcout << L"FinalResult: status:" << (int)e.Result.TranslationStatus << L". Recognized Text: " << e.Result.Text << '\n';
    for (const auto& it : e.Result.Translations)
    {
        wcout << L"    Translated into " << it.first.c_str() << ":" << it.second.c_str();
    }
}

static void OnSynthesisResult(const TranslationSynthesisResultEventArgs& e)
{
    wcout << L"Translation syntheis result: size of audio data: " << e.Result.Audio.size();
}

static void OnNoMatch(const TranslationTextResultEventArgs& e)
{
    wcout << L"NoMatch:" << (int)e.Result.Reason << '\n';
}

static void OnCanceled(const TranslationTextResultEventArgs& e)
{
    wcout << L"Canceled:" << (int)e.Result.Reason << L"- " << e.Result.Text << '\n';
}

// Continuous translation.
void TranslationContinuousRecognitionUsingEvents()
{
    // Gets recognizer factory.
    auto factory = RecognizerFactory::GetDefault();

    // Sets subscription key
    // Replaces with your own subscription key.
    factory->SetSubscriptionKey(L"YourSubscriptionKey");

    // Sets source and target languages
    auto fromLanguage = L"en-US";
    vector<wstring> toLanguages{ L"de-DE", L"fr-FR" };

    // Creates a translation recognizer using microphone as audio input.
    auto recognizer = factory->CreateTranslationRecognizer(fromLanguage, toLanguages);

    // Subscribes to events.
    recognizer->IntermediateResult.Connect(&OnPartialResult);
    recognizer->FinalResult.Connect(&OnFinalResult);
    recognizer->Canceled.Connect(&OnCanceled);
    recognizer->NoMatch.Connect(&OnNoMatch);
    recognizer->TranslationSynthesisResultEvent.Connect(&OnSynthesisResult);

    wcout << L"Say something...\n";

    // Starts continuos recognition. Uses StopContinuousRecognitionAsync() to stop recognition.
    recognizer->StartContinuousRecognitionAsync().wait();

    wcout << L"Press any key to stop\n";
    string s;
    getline(cin, s);

    // Stops recognition.
    recognizer->StopContinuousRecognitionAsync().wait();

    // Unsubscribes to events.
    recognizer->IntermediateResult.Disconnect(&OnPartialResult);
    recognizer->FinalResult.Disconnect(&OnFinalResult);
    recognizer->Canceled.Disconnect(&OnCanceled);
    recognizer->NoMatch.Disconnect(&OnNoMatch);
    recognizer->TranslationSynthesisResultEvent.Disconnect(&OnSynthesisResult);
}
// </TranslationContinuousRecognitionUsingEvents>

File: samples/csharp_samples/translation_samples.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

// <toplevel>
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.CognitiveServices.Speech.Recognition;
// </toplevel>

namespace MicrosoftSpeechSDKSamples
{
    public class TranslationSamples
    {
        // Translation from microphone.
        // <TranslationWithMicrophoneAsync>
        public static async Task TranslationWithMicrophoneAsync()
        {
            // Creates an instance of a speech factory with specified
            // subscription key.
            var factory = SpeechFactory.FromSubscription("YourSubscriptionKey", "");

            // Sets source and target languages
            string fromLanguage = "en-US";
            List<string> toLanguages = new List<string>() { "de-DE", "fr-FR" };

            // Creates a translation recognizer using microphone as audio input.
            using (var recognizer = factory.CreateTranslationRecognizer(fromLanguage, toLanguages))
            {
                Console.WriteLine("Say something...");

                // Starts translation. It returns when the first utterance has been recognized.
                var result = await recognizer.RecognizeAsync().ConfigureAwait(false);

                // Checks result.
                if (result.Reason != RecognitionStatus.Success)
                {
                    Console.WriteLine($"There was an error in speech recognition, reason {result.Reason}-{result.RecognizedText}");
                }
                else if (result.TranslationStatus != RecognitionStatus.Success)
                {
                    Console.WriteLine($"There was an error in translation, status: {result.TranslationStatus}");
                }
                else
                {
                    Console.WriteLine($"We recognized in {fromLanguage}: {result.RecognizedText}");
                    foreach (var element in result.Translations)
                    {
                        Console.WriteLine($"    Translated into {element.Key}: <{element.Value}>");
                    }
                }
            }
        }
        // </TranslationWithMicrophoneAsync>

        // Translation using file input.
        // <TranslationWithFileAsync>
        public static async Task TranslationWithFileAsync()
        {
            // Creates an instance of a speech factory with specified
            // subscription key.
            var factory = SpeechFactory.FromSubscription("YourSubscriptionKey", "");

            // Sets source and target languages
            string fromLanguage = "en-US";
            List<string> toLanguages = new List<string>() { "de-DE", "fr-FR" };

            // Creates a translation recognizer using file as audio input.
            // Replace with your own audio file name.
            using (var recognizer = factory.CreateTranslationRecognizerWithFileInput(@"YourAudioFileName", fromLanguage, toLanguages))
            {
                // Starts translation. It returns when the first utterance has been recognized.
                var result = await recognizer.RecognizeAsync().ConfigureAwait(false);

                // Checks result.
                if (result.Reason != RecognitionStatus.Success)
                {
                    Console.WriteLine($"There was an error in speech recognition, reason {result.Reason}-{result.RecognizedText}");
                }
                else if (result.TranslationStatus != RecognitionStatus.Success)
                {
                    Console.WriteLine($"There was an error in translation, status: {result.TranslationStatus}");
                }
                else
                {
                    Console.WriteLine($"We recognized in {fromLanguage}: {result.RecognizedText}");
                    foreach (var element in result.Translations)
                    {
                        Console.WriteLine($"    Translated into {element.Key}: <{element.Value}>");
                    }
                }
            }
        }
        // </TranslationWithFileAsync>

        // Translation continuous recognition with events
        // <TranslationContinuousRecognitionAsync>
        public static async Task TranslationContinuousRecognitionAsync()
        {
            // Creates an instance of a speech factory with specified
            // subscription key.
            var factory = SpeechFactory.FromSubscription("YourSubscriptionKey", "");

            // Sets source and target languages
            string fromLanguage = "en-US";
            List<string> toLanguages = new List<string>() { "de-DE", "fr-FR" };

            // Creates a translation recognizer using microphone as audio input.
            using (var recognizer = factory.CreateTranslationRecognizer(fromLanguage, toLanguages))
            {
                // Subscribes to events.
                recognizer.IntermediateResultReceived += (s, e) => {
                    Console.WriteLine($"\nPartial result: recognized in {fromLanguage}: {e.Result.RecognizedText}.");
                    foreach (var element in e.Result.Translations)
                    {
                        Console.WriteLine($"    Translated into {element.Key}: <{element.Value}>");
                    }
                };

                recognizer.FinalResultReceived += (s, e) => {
                    Console.WriteLine($"\nFinal result: Status: {e.Result.TranslationStatus}, recognized text in {fromLanguage}: {e.Result.RecognizedText}.");
                    foreach (var element in e.Result.Translations)
                    {
                        Console.WriteLine($"    Translated into {element.Key}: <{element.Value}>");
                    }
                };

                recognizer.SynthesisResultReceived += (s, e) =>
                {
                    Console.WriteLine($"Synthesis result received. Size of audio data: {e.Result.Audio.Length}");
                };

                recognizer.RecognitionErrorRaised += (s, e) => {
                    Console.WriteLine($"\nSorry, error occurred. Reason: {e.Status.ToString()}");
                };

                recognizer.OnSessionEvent += (s, e) => {
                    Console.WriteLine($"\nSession event. Event: {e.EventType.ToString()}.");
                };

                // Starts continuous recognition. Uses StopContinuousRecognitionAsync() to stop recognition.
                Console.WriteLine("Say something...");
                await recognizer.StartContinuousRecognitionAsync().ConfigureAwait(false);

                Console.WriteLine("Press any key to stop");
                Console.ReadKey();

                await recognizer.StopContinuousRecognitionAsync().ConfigureAwait(false);
            }
        }
        // </TranslationContinuousRecognitionAsync>
    }
}
